### PR TITLE
BUGFIX: Show and apply all changes in workspace review

### LINKED ISSUE #4587: BUG:  Workspace review doesn't show new documents and their child nodes
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Creating a new document node and adding content nodes will update the publish button accordingly. But the changes will not show up if the user clicks "Review changes".

### Expected Behavior

I can review all unpublished changes in the workspace module and publish them.

### Steps To Reproduce

1. Create a new page
2. Creat a new content element inside the page
3. Click "Review changes" button

### Environment

```markdown
- Flow: 9.0
- Neos: 9.0
- PHP: 8.2
```

### Anything else?

Relates: #4573

### PR DESCRIPTION
**Pull request by ahaeslich**
Fixes: #4587
Relates: #4563

**Checklist**

- [x] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

### CODE REVIEW

**Comment by ahaeslich** [comment]
While testing this change in our project we found that:

Selecting all nodes by using "Select all current changes" and click on " Publish selected changes to live" (Calls `publishOrDiscardNodesAction`) works. ✅ 

Selecting nothing and click on " Publish all changes to live" (Calls `publishWorkspaceAction`) will lead to this error message:

> The base workspace has been modified in the meantime; please rebase.

Edit: ... sometimes 🤯 
Edit 2: can reproduce this strange behaviour without my changes ... happy us ... this calls for a new issue (https://github.com/neos/neos-development-collection/issues/4591)

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php [hunk 1]
```diff
@@ -792,44 +782,60 @@ protected function computeSiteChanges(Workspace $selectedWorkspace, ContentRepos
...
+                        $siteNode = $ancestor;
                     }
                 }
 
                 // Neither $documentNode, $siteNode or its cannot really be null, this is just for type checks;
                 // We should probably throw an exception though
                 if ($documentNode !== null && $siteNode !== null && $siteNode->nodeName) {
                     $siteNodeName = $siteNode->nodeName->value;
-                    $documentPath = implode('/', array_slice(array_map(
+                    $documentPath = implode('/', array_reverse(array_map(
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:805]: I'm trying to understand: (how) is this related to the issue?

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php [hunk 2]
```diff
@@ -792,44 +782,60 @@ protected function computeSiteChanges(Workspace $selectedWorkspace, ContentRepos
...
                         fn (NodeName $nodeName): string => $nodeName->value,
                         $documentPathSegments
-                    ), 2));
+                    )));
                     $relativePath = str_replace(
                         sprintf('//%s/%s', $siteNodeName, $documentPath),
                         '',
-                        implode('/', array_map(
+                        implode('/', array_reverse(array_map(
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php]: and here?

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php [hunk 3]
```diff
@@ -792,44 +782,60 @@ protected function computeSiteChanges(Workspace $selectedWorkspace, ContentRepos
...
                     if (!isset($siteChanges[$siteNodeName]['siteNode'])) {
                         $siteChanges[$siteNodeName]['siteNode']
                             = $this->siteRepository->findOneByNodeName(SiteNodeName::fromString($siteNodeName));
                     }
+
                     $siteChanges[$siteNodeName]['documents'][$documentPath]['documentNode'] = $documentNode;
+                    if ($documentNode->equals($node)) {
+                        $siteChanges[$siteNodeName]['documents'][$documentPath]['isNew'] = $change->created;
+                        $siteChanges[$siteNodeName]['documents'][$documentPath]['isMoved'] = $change->moved;
+                    }
```
Comment (bwaidelich) [inline Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:825]: I don't understand this change either. Are those `isNew` and `isMoved` magic strings new?

**Review by bwaidelich — COMMENTED** [review]
As written on Slack, I found it hard to review this PR because it contains unrelated changes (it seems).
Also without any tests it's hard to tell what changes actually fix the bug. Maybe you could add some inline comments (in code or just as PR comments) to explain the fix?

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php (hunk 3 above)
Comment (ahaeslich) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:825]: Not exactly. They by the frontend part to visiually show the difference and also for selecting the correct combination of nodes. Thats something we should probably check again and maybe change when we update the workspace module.

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php (hunk 1 above)
Comment (ahaeslich) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:805]: In both places I didn't quite understand why we need the paths in the first place. It's also used by the current JS logic to select and group nodes by document. And in neos 8 that starts at the side node and not with the lowest node.

**Comment by ahaeslich** [comment]
> As written on Slack, I found it hard to review this PR because it contains unrelated changes (it seems). Also without any tests it's hard to tell what changes actually fix the bug. Maybe you could add some inline comments (in code or just as PR comments) to explain the fix?

Thx for your questions ❤️ I added some comments in the code as I think we might need them there for a future refactoring.

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php (hunk 3 above)
Comment (kitsunet) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:825]: no, see the removed part further down around (previously 850).

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php (hunk 2 above)
Comment (ahaeslich) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php]: Same as above.

File: Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php (hunk 1 above)
Comment (bwaidelich) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:805]: > And in neos 8 that starts at the side node and not with the lowest node

So that's why we change `array_slice` to `array_reverse` right?

Comment (ahaeslich) [reply Neos.Neos/Classes/Controller/Module/Management/WorkspacesController.php:805]: Yes

**Comment by ahaeslich** [comment]
We can update this to use `Neos.Neos:Site` when #4563 is merged

**Review by mhsdesign — APPROVED** [review]
I can confirm that the changes count doesnt reflect any new changes (the green ones) neither for user workspaces or an actual workspace:

before
image

after (this fix)
image


modifications seem to be correctly reflected already:
image

and still work after this fix.

-------

So thanks for taking care of this. I approve by testing ;)

**Comment by ahaeslich** [comment]
Thx @mhsdesign for merging ❤️. So I think we should be good to go here.

**Review by bwaidelich — APPROVED** [review]
This one is a bit hard to review (by reading). But we agreed to go for it as it definitely make things better!

Thanks for taking care!
